sequencer/src/state_signature/relay_server/stake_table_tracker.rs

```diff
@@ -0,0 +1,198 @@
+        tracing::debug!("Fetch stake table for block {block_height}");
+
+        tracing::trace!("Acquire read lock for stake table info");
+        let read_guard = self.inner.read().await;
+        let (blocks_per_epoch, epoch_start_block) =
+            if let Some(blocks_per_epoch) = read_guard.blocks_per_epoch {
+                let epoch_start_block = read_guard.epoch_start_block.unwrap();
+                tracing::trace!("Drop read lock for stake table info");
+                drop(read_guard);
+                (blocks_per_epoch, epoch_start_block)
+            } else {
+                tracing::trace!("Drop read lock for stake table info");
+                drop(read_guard);
+                tracing::trace!("Acquire write lock for stake table info");
+                let mut write_guard = self.inner.write().await;
+                if let Some(blocks_per_epoch) = write_guard.blocks_per_epoch {
+                    (blocks_per_epoch, write_guard.epoch_start_block.unwrap())
+                } else {
+                    tracing::debug!("Fetching epoch config from sequencer");
+                    let (blocks_per_epoch, epoch_start_block) =
+                        fetch_epoch_config_from_sequencer(&write_guard.sequencer_url).await?;
+                    write_guard.blocks_per_epoch.get_or_insert(blocks_per_epoch);
+                    write_guard
+                        .epoch_start_block
+                        .get_or_insert(epoch_start_block);
+                    tracing::debug!(
+                        "Fetched epoch config from sequencer: blocks_per_epoch: {}, \
+                         epoch_start_block: {}",
+                        blocks_per_epoch,
+                        epoch_start_block
+                    );
+                    tracing::trace!("Drop write lock for stake table info");
+                    drop(write_guard);
+                    (blocks_per_epoch, epoch_start_block)
+                }
+            };
+        if block_height <= epoch_start_block || blocks_per_epoch == 0 {
+            return self.genesis_stake_table_info().await;
+        }
+
+        let epoch = epoch_from_block_number(block_height, blocks_per_epoch);
+        tracing::trace!("Acquire read lock for stake table info");
+        let read_guard = self.inner.read().await;
+        if let Some(stake_table_info) = read_guard.stake_table_infos.get(&epoch) {
+            return Ok(stake_table_info.clone());
+        }
+        tracing::trace!("Drop read lock for stake table info");
+        drop(read_guard);
+        tracing::trace!("Acquire write lock for stake table info");
+        let mut write_guard = self.inner.write().await;
+        if let Some(stake_table_info) = write_guard.stake_table_infos.get(&epoch) {
+            return Ok(stake_table_info.clone());
+        }
+
+        tracing::debug!("Fetching stake table for epoch {} from sequencer", epoch);
+        let stake_table = fetch_stake_table_from_sequencer(
```

Comment:
here again, should check after obtaining the write lock and before doing network fetch.

Reply:
checked in L160